### ISSUE TITLE
NickelSeries not working on Kobo Libra2

### ISSUE THREAD

**Issue by wirdehall** [geek1011/kobo-mods#5]
I installed NickleSeries on my Kobo Libra 2 but nothing seams to happen when I transfer a new .kepub to the device.

Background:
I'm using https://github.com/cyrilis/epub-gen for generating epub3 files. But since it didn't have support for adding collections/series I forked the repo and added the functionallity and created a [PR](https://github.com/cyrilis/epub-gen/pull/110). Currently I'm using my fork in wait for PR to be merged(hopefully).
My fork is here: https://github.com/wirdehall/epub-gen

Example of metadata I'm adding with my fork:
```
<meta property="belongs-to-collection" id="collection">Zenith of Sorcery</meta>
<meta refines="#collection" property="collection-type">series</meta>
<meta refines="#collection" property="group-position">1</meta>
```
After that I use Kepubify to convert the .epub (v3) to .kepub.

I have tried to sideload books both using kobo desktop (ie. usb) and the dropbox functionallity using NickelMenu.
I have used new books on every try.
Nothing seams to work. I am only transferring the .kepub books as I want the metadata that comes with them, could that be a problem?

Lastly I tried using [Seriesmeta](https://pgaskin.net/kepubify/docs/#seriesmeta) on my already uploaded books and all my unsuccessful transfers now worked so it appears the meta-data is correct.
I don't understand what I'm doing wrong. I even tried to uninstall and re-install NS but it did not help.
Any help at this point is appreciated, I want my workflow to be as automated as possible so this mod would be amazing if I could get it to work.

Best Regards
@Wirdehall

**Comment by pgaskin** [comment]
Did any of the newly-transferred books show the correct series metadata with NickelSeries?

**Comment by wirdehall** [comment]
Hello,
Thanks for the reply!

No, that's the problem I'm having.
Only when I user Seriesmeta did it work which tells me NickelSeries issn't working on my device.

**Comment by wirdehall** [comment]
Hello again @pgaskin ,
I don't want to stress you if you are busy but is there maybe some additional information I could provide that would make it easier for us to understand the issue here?

**Comment by pgaskin** [comment]
Can you get the syslog (using the logread command over telnet or via termux) after attempting to transfer some books with NickelSeries installed?

**Comment by wirdehall** [comment]
Hello again @pgaskin,
I'm sorry for the long delay, just had a kid a few days after your response so life has been kind of crazy.
Anyways, I could not connect using telnet for some reason but I manged to dump the syslog so I hope that's good enough.
Here's to log of my attempt to side-load a book called Super Minion:
[syslog.log](https://github.com/pgaskin/kobo-mods/files/12861616/syslog.log)

**Comment by pgaskin** [comment]
Did the metadata appear after transferring the book? It seems to have worked according to the syslog. If not, does it appear in the `content` table in the database?

**Comment by wirdehall** [comment]
Okay, that's crazy!
It worked for that book. I had to try it again to see if it was fixed because I haven't seen it work for any other book previously.
So I imported a new book today, it did not show up as a series this time, super weird...
So it worked once? But now it doesn't work again?
Here's the syslog from that import:
[syslog.log](https://github.com/pgaskin/kobo-mods/files/12887513/syslog.log)

**Comment by wirdehall** [comment]
Okay more info now, tried to import 3 more books. The three of them all showed up in the Series tab indicating everything went fine.
So this appears to be the worst kind of bug, an intermittent bug. Or maybe there is something wrong with my metadata but if that were the case it's weird that the [Seriesmeta](https://pgaskin.net/kepubify/docs/#seriesmeta) script can fix the issue.

**Comment by NiLuJe** [comment]
Have you checked the integrity of the Nickel db?

Or the sanity of the filesystem after the spurious transfer?

There are (apparently) still known resource leaks that can lead to weird and funky data integrity issues in Nickel (because of the way it handles USBMS, with a lazy unmount instead of an atomic one).

The usual workaround is to: first, verify database integrity & then filesystem integrity to start from a possibly clean slate. Then start from a cold boot to confirm that. And, then, to never ever start an USBMS session from basically anything else than the Home screen. It's apparently known to be broken when launched from within a book, but given past issues, I wouldn't necessarily trust it from within a Library view after a book open either...

**Comment by pgaskin** [comment]
Might be https://github.com/pgaskin/kobo-mods/commit/0deffca74bb3c82cc260dcc373ebfcc90b3bc002, actually. I'll do a proper release with that commit in a bit.

**Comment by pgaskin** [comment]
@wirdehall try NickelSeries v4 and see if it still has the issue.

**Comment by wirdehall** [comment]
> Have you checked the integrity of the Nickel db?
> 
> Or the sanity of the filesystem after the spurious transfer?

@NiLuJe I'm not sure how to do that :)

@pgaskin I installed it, it made all my screens inverted in color and I couldn't get it to re-invert even though I added the option to NickleMenu. Had to Uninstall NickleSeries and reinstall the v4 and now it seamed to work.
I have only tired importing one new series thus far, it looks good. I will keep you posted if I find any more issues.

**Comment by wirdehall** [comment]
PS. The two first download links on https://pgaskin.net/kepubify/ns/ still links to v3. Maybe it's by design but just thought you might want to know :)

**Comment by pgaskin** [comment]
>  it made all my screens inverted in color 

It did *what*? You're saying upgrading NickelSeries inverted your screen?

> first download links on https://pgaskin.net/kepubify/ns/ still links to v3. 

Thanks.

**Comment by wirdehall** [comment]
> It did what? You're saying upgrading NickelSeries inverted your screen?

Yes indeed, I'm thinking it has something to do with my NickelMenu.
I had previously added the invert screen menu item acording to the docs here: https://github.com/pgaskin/NickelMenu/blob/master/res/doc#L204
But for some reason it didn't work at the time, I thought "oh well, no matter" and moved on with my life.
I'm guessing upgrading NS for some reason made that config work again but for some reason it did it not only for the reader portion but for the entire tablet (even in the homescreen). I could toggle the inversion inside the reader but I couldn't get it to re-invert in the homescreen so in the end I uninstalled NS v4 and then the screen went back to normal again. After that I re-installed NS v4 again and this time I had no inversion problem.

Super weird but it solved it self, just thought it was an interesting phenomenon you might be intrested in hearing about, maybe it will happen to others? Maybe it wont.

**Comment by pgaskin** [comment]
> just thought it was an interesting phenomenon you might be intrested in hearing about

Yes, I heard about it happening to a few other people, and I was utterly baffled, so I thought I'd ask you since you are a bit more aware of what you're doing, lol.

> could toggle the inversion inside the reader

So you *did* toggle it; it didn't just happen on its own when upgrading NickelSeries, right?

**Comment by wirdehall** [comment]
No worries I understand, the least I can do. :)

I would say that it did happen on it's own. Let me take it as detailed as I can chronologically to prevent any misunderstanding.

1. A few days before I created this ticket I installed NickelMenu and was playing around with it. One of the settings I tried were invert screen. It didn't work at the time, the screen just flickered so I dropped it and assumed it was a bug of some kind, not important enough for me to create a ticket over.
2. A few days later I installed NickelSeries and were experiencing the issues with imports of metadata that created this issue.
3. At this point I hadn't gotten inversion to work at all and had written it of as something I wasn't going to use. And I hadn't touched the setting for a good long while.
4. You release NSv4 and I install it. The first thing I notice after I install it when it's booted into the homescreen is that all the colors are inverted. I kind of mini-panic at this point as I'm thinking I might have f'ed something up and thought that it might be broken. I reboot a few times and try going in and out of the reader, trying to activate dark-mode in the reader which makes the reader light when dark mode is activated.
5. At this point I remember the setting that I played around with a few weeks earlier that said it was going to invert the screen. So I try to activate it, and true enough I can get it to invert inside the reader and now get the reader to behave normally, with dark mode being the dark one and light more being the light one. But at this point the homescreen and everything outside of the reader is still inverted.
6. So I go to the docs and see if there is a way to toggle inversion for the whole device, I add a menu-item for the main-menu to toggle it for :main but after testing it, it still doesn't work so I decided to try to uninstall NSv4.
7. I uninstall NSv4, boot up the device and everything works as it should again.
8. I re-install NSv4, boot up the device and everything still looks good.

I see now that I looked through the documentation again that it requires the device to reboot on newer devices so maybe it was just some bad luck with me trying this and I just happened to get it toggled correctly when I uninstalled NSv4.
https://github.com/pgaskin/NickelMenu/blob/master/res/doc#L92
But that still doesn't explain why it inverted in the first place after I installed NSv4, I hadn't played around with that settings for a few weeks when I installed NSv4 the first time.

Hope this brings some clarity as to how it all went down :)

**Comment by pgaskin** [comment]
Okay, so it seems that you toggled it on, thought it didn't work, then next rebooted to install NSv4, which allowed it to work, you tested the toggle again (an odd number of times), then rebooted to uninstall NSv4, which applied the invert setting.

So basically, you (and others) have tested the invert screen setting without realizing it now requires a reboot to apply.

I should probably make NM show a warning or something by default for invert when it requires a reboot.

**Comment by wirdehall** [comment]
Yeah that's probably what happend but I think I have rebooted the device between me playing with the setting and installing NSv4 but I can't be sure.

Let's say I didn't for our sanitys sake eh? :D

Yeah that would probably be a good idea, it would probably reduce confused support tickets a bit.
